gbpservice/neutron/nsf/core/main.py

```diff
@@ -0,0 +1,705 @@
+
+
+class EventHandlers(object):
+
+    def __init__(self):
+        self._ehs = {}
+
+    def register(self, ev):
+        LOG.debug(_("Registering handler %s" % (self.identify(ev))))
+        if ev.id in self._ehs.keys():
+            self._ehs[ev.id].extend([ev])
+        else:
+            self._ehs[ev.id] = [ev]
+
+    def get(self, ev):
+        for id, eh in self._ehs.iteritems():
+            if id == ev.id:
+                LOG.debug(_("Returning handler %s" % (self.identify(eh[0]))))
+                return eh[0].handler
+        return None
+
+    def identify(self, ev):
+        return "%s - %s" % (ev.identify(), identify(ev.handler))
+
+
+class ServiceController(object):
+
+    def __init__(self, conf, modules):
+        self._conf = conf
+        self.modules = modules
+        self._lock = Lock()
+        self._serializer = Serializer(self)
+
+    def lock(self):
+        self._lock.acquire()
+        LOG.debug(_("Acquired lock.."))
+
+    def unlock(self):
+        self._lock.release()
+        LOG.debug(_("Released lock.."))
+
+    def event_done(self, ev):
+        LOG.info(_("Event %s done" % (ev.identify())))
+        mapp = self._serializer.copy()
+        mapp = mapp[ev.worker_attached]
+        if ev.binding_key not in mapp:
+            return
+
+        LOG.debug(_("Checking if event %s in serialize Q"
+                    % (ev.identify())))
+        qu = mapp[ev.binding_key]['queue']
+        for elem in qu:
+            if elem.key == ev.key:
+                LOG.debug(_("Removing event %s from serialize Q"
+                            % (elem.identify())))
+                self._serializer.remove(elem)
+                break
+        self._serializer.deserialize(ev)
+
+    def serialize(self, ev):
+        if not ev.serialize:
+            return ev
+        if not self._serializer.serialize(ev):
+            return ev
+        return None
+
+    def serialize_get(self):
+        LOG.debug(_(""))
+        smap = self._serializer.copy()
+        for mapp in smap.values():
+            for val in mapp.values():
+                if val['in_use']:
+                    continue
+                else:
+                    if val['queue'] == []:
+                        continue
+                    ev = val['queue'][0]
+                    LOG.debug(_("Returing serialized event %s"
+                                % (ev.identify())))
+                    return ev
+        return None
+
+    def workers_init(self):
+        wc = 2 * (multiprocessing.cpu_count())
+        if cfg.CONF.workers != wc:
+            wc = cfg.CONF.workers
+            LOG.info(_("Creating %d number of workers" % (wc)))
+
+        workers = [tuple() for w in range(0, wc)]
+
+        for w in range(0, wc):
+            qu = Queue()
+            evworker = EventHandler(self, qu, self.ehs)
+            proc = Process(target=evworker.run, args=(qu,))
+            proc.daemon = True
+            workers[w] = workers[w] + (proc, qu, evworker)
```

Comment:
Should the list (proc, qu, evworker) be a dict?
workers should be event_worker_descs